File: erlang/pe010.erl
-module(pe010).
-export([start/0]).

maxv() -> 2000000.

start() ->
  pe010(maxv(), 2, 0).

pe010(MaxV, MaxP, Sum) when (MaxP >= MaxV) ->
  io:format("~w~n", [Sum]);

pe010(MaxV, MaxP, Sum) ->
  pe010(MaxV, next_prime(MaxP), Sum + MaxP).

next_prime(Number) when (Number < 2) ->
  2;

next_prime(Number) when (Number == 2) ->
  3;

next_prime(Number) when (Number rem 2 == 0) ->
  next_prime(Number - 1);

next_prime(Number) ->
  Try = Number + 2,
  IsPrime = is_prime(Try),
  if
    IsPrime ->
      Try;
    true ->
      next_prime(Try)
  end.

is_prime(Number) when ((Number == 1) or (Number == 0))  ->
  false;

is_prime(Number) when (Number == 2) ->
  true;

is_prime(Number) when (Number rem 2 == 0) ->
  false;

is_prime(Number) ->
  prime_test(3, Number).

prime_test(Div, Number) when ((Div * Div) > Number) ->
  true;

prime_test(Div, Number) when (Number rem Div == 0) ->
  false;

prime_test(Div, Number) ->
  prime_test(Div + 2, Number).

File: ocaml/pe010.ml
open Int64

let maxi = 2_000_000L ;;

let divisible x y  = 
  Int64.(rem x y = zero) ;;

let rec prime_test n maxv =
  if (mul n n) > maxv then
    true
  else
    if (divisible maxv n) then
      false
    else
      prime_test (add n 2L) maxv ;;

let is_prime n =
  match n with
  | 0L   -> false
  | 1L   -> false
  | 2L   -> true
  | _    -> if divisible n 2L then false else prime_test 3L n ;;

let rec pe010 isum count =
  if count > maxi then
    Printf.printf "%Ld\n" isum
  else
    if is_prime count then
      pe010 (add isum count) (add count 1L) 
    else
      pe010 isum (succ count) ;;

let main () =
  pe010 0L 1L ;;

let () = main ()
